### PR TITLE
SeccompError improvements

### PR DESCRIPTION
**Pull request by rusty-snake**
> **Add From impl for SeccompError**
> This will impl From<std::ffi::NulError|std::num::TryFromIntError|std::str::Utf8Error>
> for SeccompError. These impls are public, but I don't think that's a
> problem.
> 
> Unfortunately does `impl<T: Error> From<T> for SeccompError` not work in
> stable rust yet.

> **Replace map_err with implicit From::from**

### CODE REVIEW

File: libseccomp/src/error.rs [hunk 1]
```diff
@@ -64,6 +64,27 @@ impl Error for SeccompError {
...
+        Self::with_source(ErrorKind::Common(err.to_string()), err)
+    }
+}
+*/
+
+macro_rules! impl_seccomperror_from {
+    ($errty:ty) => {
+        impl From<$errty> for SeccompError {
+            fn from(err: $errty) -> Self {
+                Self::with_source(ErrorKind::Common(err.to_string()), err)
```
Comment (ManaSugi) [inline libseccomp/src/error.rs:79]: In this case, the output will be `<ErrorType> caused by: <ErrorType>` by [line 51](https://github.com/libseccomp-rs/libseccomp-rs/pull/27/files#diff-0186c07869ec66b16152774dde29b8b526290c7edd8868be341e76072f226bd7R51).
Example of the case of `NulError`:
`NulError  caused by: NulError`

Hence, I'd like to suggest `Errorkind::Common("Convert error", err)` or something like that.
I'm not sure a proper message as generic.

Comment (rusty-snake) [reply libseccomp/src/error.rs:79]: The `ToString` implementation of `NulError` (https://doc.rust-lang.org/src/std/ffi/c_str.rs.html#1048-1052) looks like

```rust
impl fmt::Display for NulError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "nul byte found in provided data at position: {}", self.0)
    }
}
```

so the actually error would look like
`nul byte found in provided data at position: 11 caused by: Err(NulError(11, [72, 101, 108, 108, 111, 32, 87, 111, 114, 108, 100, 0, 33]))`

Is there any reason why we format the source with Debug? If not, what about using Display for the source and adding a new `ErrorKind::Source`?

Comment (rusty-snake) [reply libseccomp/src/error.rs:79]: BTW what was the intention of `SeccompError.code`? It is never read. Can we remove it?

File: libseccomp/src/error.rs [hunk 2]
```diff
@@ -26,32 +27,23 @@ impl fmt::Display for SeccompError {
...
+                libc::EDOM => "Architecture specific failure",
+                libc::EACCES => "Setting the attribute with the given value is not allowed",
+                libc::EEXIST => "Failure regrading the existance of argument",
+                libc::EINVAL => "Invalid input",
+                libc::ENOMEM => "The library was unable to allocate enough memory",
+                libc::ECANCELED => "There was a system failure beyond the control of the library",
+                libc::EFAULT => "Internal libseccomp failure",
+                libc::ESRCH => "Unable to load the filter due to thread issues",
+                libc::EOPNOTSUPP => "The library doesn't support the particular operation",
+                _ => "Other failure",
```
Comment (ManaSugi) [inline libseccomp/src/error.rs:39]: Could you change the line as follows?
```rust
_=> return format!("Error caused by: errno({})", -(*e)).fmt(f),  
```

File: libseccomp/src/error.rs (hunk 1 above)
Comment (ManaSugi) [reply libseccomp/src/error.rs]: Yes, but could you leave `code` for `Debug` trait implementation.
I'll add `Debug` soon.

**Comment by rusty-snake** [comment]
I start thinking if it would make sense to move the source into `Source(Box<___>)`. i.e. a error source can only be used with Kind Source.

**Comment by rusty-snake** [comment]
`Source(Box<___>)` seems to be a bad idea.

**Comment by ManaSugi** [comment]
LGTM. Thank you!

**Comment by rusty-snake** [comment]
I think this is mostly finish.
I will rebase+squash and re-add (without `:?`)
```rust
if let Some(source) = &self.source {
    write!(f, "{} caused by: {}", msg, source)
} else {
    write!(f, "{}", msg)
}
```

**Comment by ManaSugi** [comment]
Thanks, please rebase and re-add it

**Comment by ManaSugi** [comment]
re-run ci

**Comment by ManaSugi** [comment]
LGTM, Thank you!

**Review by ManaSugi — APPROVED** [review]

**Pull request merged**